components/script_traits/lib.rs

```diff
@@ -117,10 +118,25 @@ pub enum LayoutControlMsg {
     PaintMetric(Epoch, u64),
 }
 
+/// The origin where a given load was initiated.
+/// Useful for origin checks, for example before evaluation a JS URL.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum LoadOrigin {
+    /// A load originating in the constellation.
+    TopLevel,
```

Comment:
This seems an odd name, since the constellation may be loading an iframe. Is the intension that this is for loads initiated by navigation and history traversal?

Reply:
> Is the intension that this is for loads initiated by navigation and history traversal?
No that is not the intention.

The changes in this PR would only use this variant in the following three cases:

1. When the compositor tells us to load a new url for a given existing top-level(I guess corresponding to a user manually typing a url in the address bar and hitting return).
2. When the compositor tells us to load a new top-level for a given url(I guess corresponding to a user manually opening a new tab). 
3. When a top-level panics and we need to show an error page.

Reply:
Might be worth naming those cases, e.g. `Navigation`, `Traversal`, `Initial`, `Error`? The current name `TopLevel` isn't ideal, since traversing the session history can end up loading a document inside an iframe.

Reply:
Ok so I had missed that a `LoadData` was stored on the pipeline, and potentially used to reload the page as part of traversal if it had been evicted from the bfcache. 

And in that case, would a `LoadOriging::Traversal` be appropriate? It would require us to mutate the load-data I think, since the original load isn't going to be `Traversal`, right? Also, it might make sense to use the original loadata, since we're merely reloading the page according to the original parameters, which in some way should be transparent to script(which shouldn't be aware of whether a reload was necessary or not, since it should just get the same page originally stored in the bfcache?)?

So perhaps we should add variants for the various specific cases currently all grouped under `TopLevel`, and I'm wondering if reloading because of traversal should not keep the original `LoadOrigin::Script`. 

> since traversing the session history can end up loading a document inside an iframe

Is that a special case different from the general "reloading" of a page evicted from the bfcache?

Reply:
How about `LoadOrigin::Constellation` for those cases where we actually call `LoadData::new` in the constellation, and `LoadOrigin::Script(origin)` for where we call it from script? Then a potential reload as part of history traversal will re-use the `LoadData` initially created from script, and only reloaded by the constellation, contain the original `LoadOrigin::Script(origin)`?

Also note that currently the concept is only used at https://github.com/servo/servo/pull/23368/files#diff-1879ac6bd5d567e2aa43529d33474677R862